api/src/main/java/com/debugmafia/clueless/model/Board.java

```
@@ -0,0 +1,220 @@
+          deckOCards.remove(c);
+        }
+      }
+
+      winningCards = cardsToReturn;
+    } else {
+      cardsToReturn = winningCards;
+    }
+    return cardsToReturn;
+  }
+
+  public Set<Card> draw(int numToDraw) {
+
+    HashSet<Card> setToReturn = null;
+
+    if (hasDeckBeenShuffled && deckOCards.size() >= numToDraw) {
+
+      Iterator<Card> iterator = deckOCards.iterator();
+
+      while(numToDraw > 0 && iterator.hasNext())
+      {
+        setToReturn.add(iterator.next());
+        iterator.remove();
+        numToDraw--;
+      }
+    }
+    return setToReturn;
+  }
+
+  public Set<BoardLocation> getOpenAdjacentLocations(BoardLocation locationToQuery) {
+    Set<BoardLocation> validLocations = new HashSet<BoardLocation>();
+
+    for (BoardLocation[] row : grid) {
+      for(BoardLocation column: row)
+      {
+      if (locationToQuery.getXcoord() + 1 == column.getXcoord() || locationToQuery.getXcoord() + 1 == column.getYcoord()) {
+        if (column.getType() == BoardLocationType.ROOM) {
+          validLocations.add(column);
+        } else if (column.getType() == BoardLocationType.HALLWAY && locationToQuery.getPieces().isEmpty()) {
+          validLocations.add(column);
+        }
+      } else if (locationToQuery.getSecretPassage() != null && locationToQuery.getSecretPassage() == column) {
+        validLocations.add(column);
+      }
+    }
+    }
+
+    return validLocations;
+  }
+
+  public Board movePiece(Piece p, BoardLocation to) {
+
+    BoardLocation localLocation = null;
+    Piece localPiece = null;
+
+    for(BoardLocation[] row : grid )    
+    {
+      for(BoardLocation column : row)
+      {
+        if(column == to)
+        {
+          localLocation = column;
+          break;
+        }
+      }
+    }
+
+    for(Piece pIter : gamePieces)
+    {
+      if(pIter == p)
+      {
+        localPiece = pIter;
+        break;
+      }
+    }
+
+    if(localLocation == null || localPiece == null)
+    {
+      return null;
```

If this statement executes, something bad has happened. Either the Location they're trying to move to is null, or the piece is. In either case, we should probably throw an error.